### ISSUE TITLE
Using with Jasperosft Reports or JDBC Manager

### ISSUE THREAD

**Issue by Trouble123** [Anchormen/sql4es#4]
Hi
I have been looking for a ES JDBC driver for months, so very happy i came across this.
I am trying to use it with Jaspersoft Report Studio. When i do a test, it shows it succesful, but when i go through and would normally see the tables/DB (when connecting to MS SQL), it just sits on Pending... under Table, View & Global Temporary.
I can connect to it with JDBC Manager on WIndows but get similar result where it just sits there.

Can you point me to where i can look to see where the issue could be?

Thanks

**Comment by corneversloot** [comment]
Hi,
I have noticed different applications using the driver in different ways to get database and table information which occasionally does not work. I will try to find out what is going on when using the driver from Jaspersoft and it would help if you can supply the URL you use to connect to your Elasticsearch cluster (and the version of ES you are using).
You can also try connecting with [SqlWorkbenchJ ](http://www.sql-workbench.net/downloads.html) just to check if that works

**Comment by corneversloot** [comment]
I reconstructed the error using Jaspersoft Studio 6.2.1 and was able to fix it (at least as far is I can tell).  It nows shows all fields within the selected table and enables the user to select these to be used within a report. Please download the sql4es-0.8.2.2.jar from the releases folder and let me know if that release fixed your problem.

**Comment by Trouble123** [comment]
HI
Thanks i tried to download the jar file directly from windows or linux and it was only 35-45K in size. I then did a git clone, and it was 18Mb again. 
I can now get futher, but i am seeing multiple tables called jdbc. We are pulling the data in to ES from MS SQL using jdbc, and i am guessing its because we have multiple 'pulls', and they might all be called jdbc (just copied/pasted the first one and changed the SQL query and ES index name).
Here is what i am seeing [https://dl.dropboxusercontent.com/u/79396560/share/ES.PNG](url). I dont have an issue with the fact there are multiple jdbc, but in my query what is my select statement going to be? 
When i open one of those jdbc branches in the screenshot i see heaps of lines, looks like it is actually showing all the 'fields' from all my indexies - one after another.

My JDBC URL is jdbc:sql4es://ES_SERVER_IP:9300/csrnumber? (where csrnumber is one of the ES indeies), do i a 'select * from csrnumber'

**Comment by corneversloot** [comment]
You are right regarding the URL, in your example you connect to the csrnumber index. Conceptually the driver interprets indices as databases and types within the index as its tables. In your case the query should be 'select * from csrnumber.jdbc' or alternatively just 'select * from jdbc'. You can just try that to see what you get.

However, im a bit confused by your screenshot :) That shows you have multiple jdbc types within the csrnumber index you are connected to which should not be possible. It could be there is still something wrong with the driver providing this information or how Jaspersoft Studio interprets it. It would be good to know the exact situation in ES to understand it. You can use sqlWorkbenchJ for this as i mentioned before or access ES webservice directly.

**Comment by Trouble123** [comment]
Hi, i setup sqlWorkBenchJ and can connect,and select * from csrnumber.jdbc also works but i think its getting all the records from all the tables. The reason is that i ran 'select _index, _id from xcsrcallfulldetails2.jdbc', and this only 44 documents in ES, yet it took 26 mins to run and i have 97 rsult tabs!

**Comment by corneversloot** [comment]
Could it be that the 'csrnumber' index you connect to is an Elasticsearch Alias? An alias behaves like a normal index but can span multiple other indexes (a little bit like a view in sql). This would also explain your screenshot which shows multiple jdbc types within the csrnumber index. My guess is that the tool you use to get data from MySQL creates a new index with jdbc type on each dump and adds it to the csrnumber index. This means that when you query the csrnumber index you query all the data within all those indexes! Running a 'select * from csrnumber.jdbc' will thus fetch all data you have; 97 resultsets, each containing 10000 results (by default), suggest you have around 970K records in your MySQL database (or at least accumulated it across all dumps). 

Long story short, i believe that what you see is default behaviour when using an alias. Se [this page](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html) for an explanation. If you want to build reports on all your jdbc data you can just query csrnumber.jdbc as you do now. I hope this makes sense. 

You can review all your indexes and alias within sqlWorkbenchJ by accessing the 'Database explorer' tab and open the Schema dropdown on the left (it will contain all indices and aliases on your elasticsearch cluster)
